pkg/routeagent/controllers/route/iptables.go

```diff
@@ -45,48 +44,65 @@ func (r *Controller) createIPTableChains() error {
 
 	klog.V(4).Infof("Insert rule to allow traffic over %s interface in FORWARDing Chain", VxLANIface)
 	ruleSpec = []string{"-o", VxLANIface, "-j", "ACCEPT"}
-	if err = r.insertUnique(ipt, "filter", "FORWARD", 1, ruleSpec); err != nil {
+	if err = r.prependUnique(ipt, "filter", "FORWARD", ruleSpec); err != nil {
 		klog.Errorf("Unable to insert iptable rule in filter table to allow vxlan traffic: %v", err)
 	}
 
 	return nil
 }
 
-func (r *Controller) programIptableRulesForInterClusterTraffic(remoteCidrBlock string) {
+func (r *Controller) programIptableRulesForInterClusterTraffic(remoteCidrBlock string) error {
 	ipt, err := iptables.New()
 	if err != nil {
-		klog.Errorf("error while initializing iptables: %v", err)
+		return fmt.Errorf("error initializing iptables: %v", err)
 	}
 
 	for _, localClusterCidr := range r.localClusterCidr {
 		ruleSpec := []string{"-s", localClusterCidr, "-d", remoteCidrBlock, "-j", "ACCEPT"}
 		klog.V(4).Infof("Installing iptables rule for outgoing traffic: %s", strings.Join(ruleSpec, " "))
 		if err = ipt.AppendUnique("nat", SmPostRoutingChain, ruleSpec...); err != nil {
-			klog.Errorf("error appending iptables rule \"%s\": %v\n", strings.Join(ruleSpec, " "), err)
+			return fmt.Errorf("error appending iptables rule \"%s\": %v\n", strings.Join(ruleSpec, " "), err)
 		}
 
 		// TODO: revisit, we only have to program rules to allow traffic from the podCidr
 		ruleSpec = []string{"-s", remoteCidrBlock, "-d", localClusterCidr, "-j", "ACCEPT"}
 		klog.V(4).Infof("Installing iptables rule for incoming traffic: %s", strings.Join(ruleSpec, " "))
 		if err = ipt.AppendUnique("nat", SmPostRoutingChain, ruleSpec...); err != nil {
-			klog.Errorf("error appending iptables rule \"%s\": %v\n", strings.Join(ruleSpec, " "), err)
+			return fmt.Errorf("error appending iptables rule \"%s\": %v\n", strings.Join(ruleSpec, " "), err)
 		}
 	}
+	return nil
 }
 
-func (r *Controller) insertUnique(ipt *iptables.IPTables, table string, chain string, position int, ruleSpec []string) error {
+func (r *Controller) prependUnique(ipt *iptables.IPTables, table string, chain string, ruleSpec []string) error {
 	rules, err := ipt.List(table, chain)
 	if err != nil {
 		return fmt.Errorf("error listing the rules in %s chain: %v", chain, err)
 	}
 
-	if strings.Contains(rules[position], strings.Join(ruleSpec, " ")) {
+	if strings.Contains(rules[1], strings.Join(ruleSpec, " ")) {
 		klog.V(4).Infof("In %s table, iptables rule \"%s\", already exists.", table, strings.Join(ruleSpec, " "))
 		return nil
 	} else {
-		if err = ipt.Insert(table, chain, position, ruleSpec...); err != nil {
-			klog.Errorf("In %s table, unable to insert iptables rule \"%s\": %v\n", table, strings.Join(ruleSpec, " "), err)
+		if err = ipt.Insert(table, chain, 1, ruleSpec...); err != nil {
+			return err
```

Comment:
This can still have the issue I commented about:

Initial state

<RULE A>

You insert:

<MY RULE>
<RULE A> 

Somebody comes and inserts:

<ANOTHER RULE>
<MY RULE>
<RULE A>

You run the function next time (service reboot):
<MY RULE>
<ANOTHER RULE>
<MY RULE>
<RULE A>

unless I'm missing something

Reply:
Done